File: dsperse/src/slice/tensor_graph.py
"""
TensorGraph: Complete dependency mapping for ONNX models.

Builds a graph of tensor producers and consumers to properly handle
skip connections and cross-slice dependencies during slicing and inference.
"""
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional
import onnx


@dataclass
class TensorInfo:
    name: str
    producer_node: Optional[str] = None
    producer_node_idx: Optional[int] = None
    consumers: list[tuple[str, int]] = field(default_factory=list)
    is_input: bool = False
    is_output: bool = False
    is_initializer: bool = False


class TensorGraph:
    def __init__(self, model: onnx.ModelProto | str | Path):
        if isinstance(model, (str, Path)):
            self.model = onnx.load(str(model))
        else:
            self.model = model
        self.graph = self.model.graph

        self.tensors: dict[str, TensorInfo] = {}
        self.node_to_idx: dict[str, int] = {}
        self.idx_to_node: dict[int, str] = {}

        self._build_graph()

    def _build_graph(self):
        for inp in self.graph.input:
            self.tensors[inp.name] = TensorInfo(
                name=inp.name,
                is_input=True
            )

        for init in self.graph.initializer:
            if init.name in self.tensors:
                self.tensors[init.name].is_initializer = True
            else:
                self.tensors[init.name] = TensorInfo(
                    name=init.name,
                    is_initializer=True
                )

        for idx, node in enumerate(self.graph.node):
            node_name = node.name or f"{node.op_type}_{idx}"
            self.node_to_idx[node_name] = idx
            self.idx_to_node[idx] = node_name

            for output in node.output:
                if output:
                    if output not in self.tensors:
                        self.tensors[output] = TensorInfo(name=output)
                    self.tensors[output].producer_node = node_name
                    self.tensors[output].producer_node_idx = idx

            for inp in node.input:
                if inp:
                    if inp not in self.tensors:
                        self.tensors[inp] = TensorInfo(name=inp)
                    self.tensors[inp].consumers.append((node_name, idx))

        for out in self.graph.output:
            if out.name in self.tensors:
                self.tensors[out.name].is_output = True

    def get_slice_inputs(self, start_idx: int, end_idx: int) -> list[str]:
        """Get all external inputs needed by nodes in range [start_idx, end_idx)."""
        internal_outputs = set()
        for idx in range(start_idx, end_idx):
            node = self.graph.node[idx]
            for output in node.output:
                if output:
                    internal_outputs.add(output)

        external_inputs = []
        seen = set()
        for idx in range(start_idx, end_idx):
            node = self.graph.node[idx]
            for inp in node.input:
                if inp and inp not in internal_outputs and inp not in seen:
                    info = self.tensors.get(inp)
                    if info and not info.is_initializer:
                        external_inputs.append(inp)
                        seen.add(inp)

        return external_inputs

    def get_slice_outputs(self, start_idx: int, end_idx: int) -> list[str]:
        """Get outputs from slice that are used by later nodes or are model outputs."""
        outputs = []
        for idx in range(start_idx, end_idx):
            node = self.graph.node[idx]
            for output in node.output:
                if not output:
                    continue
                info = self.tensors.get(output)
                if not info:
                    continue

                if info.is_output:
                    outputs.append(output)
                    continue

                for consumer_name, consumer_idx in info.consumers:
                    if consumer_idx >= end_idx:
                        outputs.append(output)
                        break

        return outputs

    def get_skip_connections(self, slice_ranges: list[tuple[int, int]]) -> dict[int, list[str]]:
        """
        For each slice, find tensors that skip over it (produced before, consumed after).

        Returns: {slice_idx: [tensor_names that skip over this slice]}
        """
        skip_connections = {}

        for slice_idx, (start, end) in enumerate(slice_ranges):
            skips = []
            for tensor_name, info in self.tensors.items():
                if info.is_initializer or info.producer_node_idx is None:
                    continue

                produced_before = info.producer_node_idx < start

                consumed_after = any(
                    consumer_idx >= end
                    for _, consumer_idx in info.consumers
                )

                if produced_before and consumed_after:
                    skips.append(tensor_name)

            skip_connections[slice_idx] = skips

        return skip_connections

    def get_tensor_lifetime(self, tensor_name: str) -> tuple[int, int]:
        """Get (first_use_idx, last_use_idx) for a tensor."""
        info = self.tensors.get(tensor_name)
        if not info:
            return (-1, -1)

        first_use = info.producer_node_idx if info.producer_node_idx is not None else 0

        if info.consumers:
            last_use = max(idx for _, idx in info.consumers)
        else:
            last_use = first_use

        if info.is_output:
            last_use = len(self.graph.node)

        return (first_use, last_use)

    def get_tensors_needed_at(self, node_idx: int) -> list[str]:
        """Get all non-initializer tensors needed as input at a specific node."""
        if node_idx >= len(self.graph.node):
            return []

        node = self.graph.node[node_idx]
        needed = []
        for inp in node.input:
            if inp:
                info = self.tensors.get(inp)
                if info and not info.is_initializer:
                    needed.append(inp)
        return needed

    def get_tensors_alive_at(self, node_idx: int) -> list[str]:
        """Get all tensors that must be kept in memory at a specific node index."""
        alive = []
        for tensor_name, info in self.tensors.items():
            if info.is_initializer:
                continue
            first, last = self.get_tensor_lifetime(tensor_name)
            if first <= node_idx <= last:
                alive.append(tensor_name)
        return alive

    def build_slice_dependencies(self, slice_ranges: list[tuple[int, int]]) -> dict:
        """
        Build complete dependency info for a set of slices.

        Returns: {
            slice_idx: {
                "inputs": [tensor_names needed from outside],
                "outputs": [tensor_names produced for later slices],
                "skip_inputs": [tensor_names from non-adjacent earlier slices],
            }
        }
        """
        result = {}

        for slice_idx, (start, end) in enumerate(slice_ranges):
            inputs = self.get_slice_inputs(start, end)
            outputs = self.get_slice_outputs(start, end)

            skip_inputs = []
            if slice_idx > 0:
                prev_end = slice_ranges[slice_idx - 1][1]
                for inp in inputs:
                    info = self.tensors.get(inp)
                    if info and info.producer_node_idx is not None:
                        if info.producer_node_idx < prev_end - 1:
                            skip_inputs.append(inp)

            result[slice_idx] = {
                "inputs": inputs,
                "outputs": outputs,
                "skip_inputs": skip_inputs,
                "node_range": (start, end),
            }

        return result

    def __repr__(self):
        return f"TensorGraph({len(self.tensors)} tensors, {len(self.graph.node)} nodes)"

File: crates/dsperse/src/slicer/tensor_graph.rs
use std::collections::{HashMap, HashSet};

use super::onnx_proto::GraphProto;

pub struct TensorInfo {
    pub name: String,
    pub producer_node: Option<String>,
    pub producer_node_idx: Option<usize>,
    pub consumers: Vec<(String, usize)>,
    pub is_input: bool,
    pub is_output: bool,
    pub is_initializer: bool,
}

pub struct TensorGraph {
    pub tensors: HashMap<String, TensorInfo>,
    pub node_to_idx: HashMap<String, usize>,
    pub idx_to_node: HashMap<usize, String>,
    num_nodes: usize,
}

impl TensorGraph {
    pub fn new(graph: &GraphProto) -> Self {
        let mut tg = Self {
            tensors: HashMap::new(),
            node_to_idx: HashMap::new(),
            idx_to_node: HashMap::new(),
            num_nodes: graph.node.len(),
        };
        tg.build(graph);
        tg
    }

    fn build(&mut self, graph: &GraphProto) {
        for inp in &graph.input {
            self.tensors.insert(
                inp.name.clone(),
                TensorInfo {
                    name: inp.name.clone(),
                    producer_node: None,
                    producer_node_idx: None,
                    consumers: Vec::new(),
                    is_input: true,
                    is_output: false,
                    is_initializer: false,
                },
            );
        }

        for init in &graph.initializer {
            if let Some(t) = self.tensors.get_mut(&init.name) {
                t.is_initializer = true;
            } else {
                self.tensors.insert(
                    init.name.clone(),
                    TensorInfo {
                        name: init.name.clone(),
                        producer_node: None,
                        producer_node_idx: None,
                        consumers: Vec::new(),
                        is_input: false,
                        is_output: false,
                        is_initializer: true,
                    },
                );
            }
        }

        for (idx, node) in graph.node.iter().enumerate() {
            let node_name = if node.name.is_empty() {
                format!("{}_{}", node.op_type, idx)
            } else {
                node.name.clone()
            };
            self.node_to_idx.insert(node_name.clone(), idx);
            self.idx_to_node.insert(idx, node_name.clone());

            for output in &node.output {
                if !output.is_empty() {
                    let t = self.tensors.entry(output.clone()).or_insert_with(|| TensorInfo {
                        name: output.clone(),
                        producer_node: None,
                        producer_node_idx: None,
                        consumers: Vec::new(),
                        is_input: false,
                        is_output: false,
                        is_initializer: false,
                    });
                    t.producer_node = Some(node_name.clone());
                    t.producer_node_idx = Some(idx);
                }
            }

            for inp in &node.input {
                if !inp.is_empty() {
                    let t = self.tensors.entry(inp.clone()).or_insert_with(|| TensorInfo {
                        name: inp.clone(),
                        producer_node: None,
                        producer_node_idx: None,
                        consumers: Vec::new(),
                        is_input: false,
                        is_output: false,
                        is_initializer: false,
                    });
                    t.consumers.push((node_name.clone(), idx));
                }
            }
        }

        for out in &graph.output {
            if let Some(t) = self.tensors.get_mut(&out.name) {
                t.is_output = true;
            }
        }
    }

    pub fn get_slice_inputs(&self, graph: &GraphProto, start_idx: usize, end_idx: usize) -> Vec<String> {
        let mut internal_outputs: HashSet<String> = HashSet::new();
        for idx in start_idx..end_idx {
            if let Some(node) = graph.node.get(idx) {
                for output in &node.output {
                    if !output.is_empty() {
                        internal_outputs.insert(output.clone());
                    }
                }
            }
        }

        let mut external_inputs = Vec::new();
        let mut seen = HashSet::new();
        for idx in start_idx..end_idx {
            if let Some(node) = graph.node.get(idx) {
                for inp in &node.input {
                    if !inp.is_empty() && !internal_outputs.contains(inp) && seen.insert(inp.clone()) {
                        if let Some(info) = self.tensors.get(inp) {
                            if !info.is_initializer {
                                external_inputs.push(inp.clone());
                            }
                        }
                    }
                }
            }
        }
        external_inputs
    }

    pub fn get_slice_outputs(&self, graph: &GraphProto, start_idx: usize, end_idx: usize) -> Vec<String> {
        let mut outputs = Vec::new();
        for idx in start_idx..end_idx {
            if let Some(node) = graph.node.get(idx) {
                for output in &node.output {
                    if output.is_empty() {
                        continue;
                    }
                    if let Some(info) = self.tensors.get(output) {
                        if info.is_output {
                            outputs.push(output.clone());
                            continue;
                        }
                        for (_consumer_name, consumer_idx) in &info.consumers {
                            if *consumer_idx >= end_idx {
                                outputs.push(output.clone());
                                break;
                            }
                        }
                    }
                }
            }
        }
        outputs
    }

    pub fn build_slice_dependencies(
        &self,
        graph: &GraphProto,
        slice_ranges: &[(usize, usize)],
    ) -> HashMap<usize, SliceDependencyInfo> {
        let mut result = HashMap::new();
        for (slice_idx, &(start, end)) in slice_ranges.iter().enumerate() {
            let inputs = self.get_slice_inputs(graph, start, end);
            let outputs = self.get_slice_outputs(graph, start, end);

            let mut skip_inputs = Vec::new();
            if slice_idx > 0 {
                let prev_end = slice_ranges[slice_idx - 1].1;
                for inp in &inputs {
                    if let Some(info) = self.tensors.get(inp) {
                        if let Some(prod_idx) = info.producer_node_idx {
                            if prod_idx < prev_end.saturating_sub(1) {
                                skip_inputs.push(inp.clone());
                            }
                        }
                    }
                }
            }

            result.insert(
                slice_idx,
                SliceDependencyInfo {
                    inputs,
                    outputs,
                    skip_inputs,
                    node_range: (start, end),
                },
            );
        }
        result
    }
}

impl std::fmt::Display for TensorGraph {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "TensorGraph({} tensors, {} nodes)",
            self.tensors.len(),
            self.num_nodes
        )
    }
}

pub struct SliceDependencyInfo {
    pub inputs: Vec<String>,
    pub outputs: Vec<String>,
    pub skip_inputs: Vec<String>,
    pub node_range: (usize, usize),
}
